File: messages/BidExchangerObjectInfo.py
from CustomDataWrapper import CustomDataWrapper
from .ObjectEffectInteger import ObjectEffectInteger
from Prints import eprint

class BidExchangerObjectInfo:
    def __init__(self) -> None:
        self.objectUID = 0
        self.objectGID = 0
        self.objectType = 0
        self.effects = []
        self.prices = []
      
    def deserialize(self, input: CustomDataWrapper):
        _id4 = 0
        _item4 = None
        _val5 = 0
        self._objectUIDFunc(input)
        self._objectGIDFunc(input)
        self._objectTypeFunc(input)
        _effectsLen = input.readUnsignedShort()

        for _ in range(_effectsLen):
            _id4 = input.readUnsignedShort()
            if _id4 == 3930:
                _item4 = ObjectEffectInteger()
                _item4.deserialize(input)
                self.effects.append(_item4)
            else:
                eprint("NOO ID4: ", _id4)

        _pricesLen = input.readUnsignedShort()
        for _ in range(_pricesLen):
            _val5 = input.readVarUhLong() # or readDouble
            if _val5 < 0 or _val5 > 9007199254740992:
                raise ValueError("Forbidden value (" + str(_val5) + ") on elements of prices.")
            self.prices.append(_val5)

    def _objectUIDFunc(self, input: CustomDataWrapper):
        self.objectUID = input.readVarUhInt()
        if self.objectUID < 0:
            raise ValueError("Forbidden value (" + str(self.objectUID) + ") on element of BidExchangerObjectInfo.objectUID.")
    
    def _objectGIDFunc(self, input: CustomDataWrapper):
        self.objectGID = input.readVarUhInt()
        if self.objectGID < 0:
            raise ValueError("Forbidden value (" + str(self.objectGID) + ") on element of BidExchangerObjectInfo.objectGID.")

    def _objectTypeFunc(self, input: CustomDataWrapper):
        self.objectType = input.readInt()
        if self.objectType < 0:
            raise ValueError("Forbidden value (" + str(self.objectType) + ") on element of BidExchangerObjectInfo.objectType.")
        
    def __str__(self):
        return f"objectUID: {self.objectUID}, objectGID: {self.objectGID}, objectType: {self.objectType}, effects: {[str(i) for i in self.effects]}, prices: {self.prices}"

File: game/network/messages/utils/BidExchangerObjectInfo.go
package game

import (
	"errors"
	"fmt"
	"sniffsniff/utils"
)

type BidExchangerObjectInfo struct {
	ObjectUID  uint32
	ObjectGID  uint32
	ObjectType uint32
	Effects    []IObjectEffect
	Prices     []uint64
}

func (message *BidExchangerObjectInfo) Deserialize(buffer *utils.Buffer) error {
	message._objectUIDFunc(buffer)
	message._objectGIDFunc(buffer)
	message._objectTypeFunc(buffer)
	numberOfEffects, err := buffer.ReadUShort()

	if err != nil {
		return err
	}
	message.Effects = make([]IObjectEffect, numberOfEffects)
	for i := 0; i < int(numberOfEffects); i++ {
		message.Effects[i], err = message.ReadEffect(buffer)
		if err != nil {
			return err
		}
	}
	numberOfPrices, err := buffer.ReadUShort()
	if err != nil {
		return err
	}
	message.Prices = make([]uint64, numberOfPrices)
	for i := 0; i < int(numberOfPrices); i++ {
		message.Prices[i], err = message.ReadPrice(buffer)
		if err != nil {
			return err
		}
	}
	return nil
}

func (message *BidExchangerObjectInfo) ReadEffect(buffer *utils.Buffer) (IObjectEffect, error) {
	id, err := buffer.ReadUShort()
	if err != nil {
		return &ObjectEffect{}, err
	}
	if id == 3930 {
		effect := ObjectEffectInteger{}
		effect.Deserialize(buffer)
		return &effect, nil
	}
	fmt.Println("Unknown type of effect " + utils.UShortToString(id))
	return &ObjectEffect{}, errors.New("Unknown type of effect " + utils.UShortToString(id))
}

func (message *BidExchangerObjectInfo) ReadPrice(buffer *utils.Buffer) (uint64, error) {
	price, err := buffer.ReadVarUhLong()
	if err != nil {
		return 0, err
	}
	if price > 9_007_199_254_740_992 {
		return 0, errors.New("Forbidden value (" + utils.ULongToString(price) + ") on elements of prices.")
	}
	return price, nil
}

func (message *BidExchangerObjectInfo) _objectUIDFunc(buffer *utils.Buffer) error {
	objectUID, err := buffer.ReadVarUhInt()
	if err != nil {
		return err
	}
	message.ObjectUID = objectUID
	return nil
}

func (message *BidExchangerObjectInfo) _objectGIDFunc(buffer *utils.Buffer) error {
	objectGID, err := buffer.ReadVarUhInt()
	if err != nil {
		return err
	}
	message.ObjectGID = objectGID
	return nil
}

func (message *BidExchangerObjectInfo) _objectTypeFunc(buffer *utils.Buffer) error {
	objectType, err := buffer.ReadUInt()
	if err != nil {
		return err
	}
	message.ObjectType = objectType
	return nil
}
